### PR TITLE
PR: ready to be merged - Lifecycle java integ test

### PR DESCRIPTION
**Pull request by maggiehe00**
This PR is to add test to 
1. Create Operator1 and domainOnPVUsingWLST with admin server and 1 managed server if they are not running. 
2. After verifying the domain is created properly, change some properties on domain
resources that would cause servers to be restarted and verify that servers are indeed restarted.
3. The properties are being tested here are: 
 env: "-Dweblogic.StdoutDebugEnabled=false"-->
 "-Dweblogic.StdoutDebugEnabled=false" 
 logHomeEnabled: true --> logHomeEnabled: false
 includeServerOutInPodLog: true --> includeServerOutInPodLog: false
 imagePullPolicy: IfNotPresent -->  imagePullPolicy: Never image: "store/oracle/weblogic:12.2.1.3" -- >image: "store/oracle/weblogic:duplicate"

After removing hard sleep, all the ITPodsRestart tests passed at Jenkins job: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/1349/

Only failed test is known issue: ITOperator.testDomainOnPVUsingWDT:131->BaseTest.testWLDFScaling

### CODE REVIEW

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITOperatorLifecycle.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+  private static String operator1File = "operator1.yaml";
+  private static String operator2File = "operator2.yaml";
+  private static final String operator_bcFile = "operator_bc.yaml";
+  private static final String operator_chainFile = "operator_chain.yaml";
+
+  // file used to customize domain properties for domain, PV and LB inputs yaml
+  private static String domainonpvwlstFile = "domainonpvwlst.yaml";
+  private static String domainonpvwdtFile = "domainonpvwdt.yaml";
+  private static String domainadminonlyFile = "domainadminonly.yaml";
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITOperatorLifecycle.java]: keep only the var members used in the code, delete the rest.

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java [hunk 2]
```diff
@@ -1034,4 +1051,62 @@ private static void checkCmdInLoopForDelete(String cmd, String matchStr, String
...
+      }
+      isLineChanged = false;
+    }
+    reader.close();
+    // writing to the file
+    Files.write(Paths.get(generatedYamlFile), changedLines.toString().getBytes());
+    logger.info("Done - generate the new yaml file ");
+  }
+
+  public static boolean findFileContainString(String inputYamlFile, String searchString)
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: Is this method to find a string in file? If so, you can use 
String content = new String ( Files.readAllBytes( Paths.get(filePath) ) );
content.indexOf(searchString) >= 0;

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITOperatorLifecycle.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+   * resources that would cause servers to be restarted and verify that servers are indeed restarted.
+   * The properties tested here are: env: "-Dweblogic.StdoutDebugEnabled=false"-->
+   * "-Dweblogic.StdoutDebugEnabled=false" logHomeEnabled: true --> logHomeEnabled: false
+   * includeServerOutInPodLog: true --> includeServerOutInPodLog: false
+   * imagePullPolicy: IfNotPresent -->  imagePullPolicy: Never
+   *
+   * @throws Exception
+   */
+  @Test
+  public void testServerRestartByDomainOnPVUsingWLST() throws Exception {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITOperatorLifecycle.java]: change test name to testServerRestartByChangingDomainProperty or something which corresponds to what is being tested?

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java [hunk 4]
```diff
@@ -850,6 +850,90 @@ private void createPV() throws Exception {
...
+  public void verifyAdminServerRestarted() throws Exception {
+    logger.info("Checking if admin pod(" + domainUid + "-" + adminServerName + ") is Terminating");
+    TestUtils.checkPodTerminating(domainUid + "-" + adminServerName, domainNS);
+    Thread.sleep(10 * 1000);
+    logger.info("Checking if admin pod(" + domainUid + "-" + adminServerName + ") is Running");
+    TestUtils.checkPodCreated(domainUid + "-" + adminServerName, domainNS);
+    Thread.sleep(10 * 1000);
+  }
+
+  public void verifyManagedServerRestarted() throws Exception {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: change method name to verifyManagedServersRestarted as its making sure all managed servers are restarted

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java (hunk 2 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: This is short and sweet :) Thanks!

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java [hunk 5]
```diff
@@ -1050,4 +1065,76 @@ private static void checkCmdInLoopForDelete(String cmd, String matchStr, String
...
+
+    Files.copy(new File(fromFile).toPath(), Paths.get(toFile), StandardCopyOption.REPLACE_EXISTING);
+  }
+
+  public static void dockerTagImage(String sourceImage, String targetImage) throws Exception {
+    logger.info("Tagging souceImage:  " + sourceImage + "  to " + targetImage);
+    String dockerCmd = "docker tag " + sourceImage + " " + targetImage;
+    logger.info("Executing cmd " + dockerCmd);
+
+    ExecResult result = ExecCommand.exec(dockerCmd);
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: See if you can replace the below lines with TestUtils.exec()

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java [hunk 6]
```diff
@@ -1050,4 +1065,76 @@ private static void checkCmdInLoopForDelete(String cmd, String matchStr, String
...
+              + result.stderr());
+    }
+  }
+
+  public static void dockerRemoveImage(String imageName) throws Exception {
+    logger.info("Removing image:  " + imageName);
+    String dockerCmd = "docker rmi -f  " + imageName;
+    logger.info("Executing cmd " + dockerCmd);
+
+    ExecResult result = ExecCommand.exec(dockerCmd);
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: See if you can replace the below lines with TestUtils.exec()

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java (hunk 5 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: Probably not. TestUtils.ExecResult exec(String cmd) and I need 2 parameters

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java (hunk 6 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: Probably not. TestUtils.ExecResult exec(String cmd) and I need 2 parameters

Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: Took another look. Gotcha :) Thanks!

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java (hunk 5 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: Took another look. Gotcha :) Thanks!

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java (hunk 6 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: Code is updated to address the above comment

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+    logger.info("SUCCESS - " + testMethodName);
+  }
+
+  private static Domain createPodsRestartdomain() throws Exception {
+
+    Map<String, Object> domainMap = TestUtils.loadYaml(DOMAINONPV_WLST_YAML);
+    domainMap.put("domainUID", "domainpodsrestart");
+    domainMap.put("adminNodePort", new Integer("30707"));
+    domainMap.put("t3ChannelPort", new Integer("30081"));
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java]: We do not need set  adminNodePort or t3Channel in this usecase

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+      logger.info("SUCCESS");
+    }
+  }
+
+  /**
+   * The property tested is: env: "-Dweblogic.StdoutDebugEnabled=false"-->
+   * "-Dweblogic.StdoutDebugEnabled=true"
+   *
+   * @throws Exception
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java]: The Javadoc should describe the usecase more detail as follows  ...

  Modify the Domain Scoped env property in the Domain Object using  kubectl apply -f domain.yaml 
  Make sure all the Server Pods in the domain got re-started 
  Reference https://github.com/oracle/weblogic-kubernetes-operator/blob/master/site/server-lifecycle.md

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java [hunk 9]
```diff
@@ -850,6 +850,92 @@ private void createPV() throws Exception {
     new PersistentVolume("/scratch/acceptance_test_pv/persistentVolume-" + domainUid, pvMap);
   }
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: Need description of the method in javadoc style.

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java [hunk 10]
```diff
@@ -850,6 +850,92 @@ private void createPV() throws Exception {
...
+      verifyAdminServerRestarted();
+      verifyManagedServersRestarted();
+      // make domain.yaml include the new changed property
+      TestUtils.copyFile(
+          BaseTest.getUserProjectsDir() + "/weblogic-domains/" + domainUid + "/domain_new.yaml",
+          BaseTest.getUserProjectsDir() + "/weblogic-domains/" + domainUid + "/domain.yaml");
+    }
+    logger.info("Done - testDomainServerPodRestart");
+  }
+
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: Need method description

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java [hunk 11]
```diff
@@ -850,6 +850,92 @@ private void createPV() throws Exception {
...
+  public void verifyAdminServerRestarted() throws Exception {
+    logger.info("Checking if admin pod(" + domainUid + "-" + adminServerName + ") is Terminating");
+    TestUtils.checkPodTerminating(domainUid + "-" + adminServerName, domainNS);
+    Thread.sleep(10 * 1000);
+    logger.info("Checking if admin pod(" + domainUid + "-" + adminServerName + ") is Running");
+    TestUtils.checkPodCreated(domainUid + "-" + adminServerName, domainNS);
+    Thread.sleep(10 * 1000);
+  }
+
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: Need Mehod Description

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java [hunk 12]
```diff
@@ -1050,4 +1065,58 @@ private static void checkCmdInLoopForDelete(String cmd, String matchStr, String
...
+    Files.write(Paths.get(generatedYamlFile), changedLines.toString().getBytes());
+    logger.info("Done - generate the new yaml file ");
+  }
+
+  public static void copyFile(String fromFile, String toFile) throws Exception {
+    logger.info("Copying file from  " + fromFile + " to " + toFile);
+
+    Files.copy(new File(fromFile).toPath(), Paths.get(toFile), StandardCopyOption.REPLACE_EXISTING);
+  }
+
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/TestUtils.java]: Need method description

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+    String testMethodName = new Object() {}.getClass().getEnclosingMethod().getName();
+    logTestBegin(testMethodName);
+
+    boolean testCompletedSuccessfully = false;
+    try {
+      logger.info(
+          "About to testDomainServerPodRestart for Domain: "
+              + domain.getDomainUid()
+              + "  Image property: store/oracle/weblogic:12.2.1.3 to store/oracle/weblogic:duplicate");
+      TestUtils.dockerTagImage("store/oracle/weblogic:12.2.1.3", "store/oracle/weblogic:duplicate");
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java]: you can replace TestUtils.dockerTagImage with TestUtils.exec("docker tag store/oracle/weblogic:12.2.1.3 store/oracle/weblogic:duplicate") and get rid of dockerTag methods.

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+      domain.testDomainServerPodRestart(
+          "\"store/oracle/weblogic:12.2.1.3\"", "\"store/oracle/weblogic:duplicate\"");
+    } finally {
+      TestUtils.dockerRemoveImage("store/oracle/weblogic:duplicate");
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java]: replace with TestUtils.exec("docker rmi -f store/oracle/weblogic:duplicate")

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java (hunk 8 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java]: Done

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java (hunk 11 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: Done

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java (hunk 10 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: Done

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java (hunk 9 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: Done

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java (hunk 7 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java]: Done

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java (hunk 13 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java]: Almost missed it. Done

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java (hunk 14 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ITPodsRestart.java]: Almost missed it. Done

**Comment by maggiehe00** [comment]
All the ITPodsRestart tests passed at Jenkins job: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/1335/

Only failed test is known issue: ITOperator.testDomainOnPVUsingWDT:131->BaseTest.testWLDFScaling

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java [hunk 15]
```diff
@@ -850,6 +850,109 @@ private void createPV() throws Exception {
...
+      for (int i = 1; i <= initialManagedServerReplicas; i++) {
+        logger.info(
+            "Checking if managed pod("
+                + domainUid
+                + "-"
+                + managedServerNameBase
+                + i
+                + ") is Terminating");
+        TestUtils.checkPodTerminating(domainUid + "-" + managedServerNameBase + i, domainNS);
+        Thread.sleep(10 * 1000);
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: TestUtils.checkPodCreated() checks in loop. Is the hard sleep needed here?

**Comment by anpanigr** [comment]
I am OK with Change based on Jenkin Run 
http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/1335/

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java (hunk 15 above)
Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: Yes. This is needed.  Between pod Terminating and Running there was a very short period "containerCreating".

Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: Hard sleep is removed.

Comment (maggiehe00) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/Domain.java]: Locally all the tests passed.

**Comment by maggiehe00** [comment]
After removing hard sleep, all the ITPodsRestart tests passed at Jenkins job: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/1349/

Only failed test is known issue: ITOperator.testDomainOnPVUsingWDT:131->BaseTest.testWLDFScaling

**Pull request merged**